### DISCUSSION TITLE
What's the difference between `o3.Linear` and `o3.nn.FullyConnectedNet`?

### DISCUSSION THREAD

**Discussion by thu-wangz17** [e3nn/e3nn#342] [Q&A] [answered]
Hi, this is an amazing work! I'm confused about the implement of SE(3)-Transformers with `e3nn`.
In the original paper of SE(3)-Transformers, it said that "The linear embedding matrices $W_Q^{lk}$ and $W^{lk}_K(xj − xi) are of TFN type (c.f. Eq. (8)). " I noticed that the implement of SE3-Transformers in the `e3nn` document uses `o3.Linear` for query and `o3.nn.FullyConnectedNet` for key and value, respectively. Thus I have the question w.r.t the difference between `o3.Linear` and `o3.nn.FullyConnectedNet`. Thank you!

**Comment by mariogeiger** [comment, marked as answer]
`o3.Linear` is of TFN type.

For key and values we use a `o3. FullyConnectedTensorProduct` and we feed its weights with a `nn. FullyConnectedNet`. 

Just to be sure we are talking about the same thing, you are looking at (the transformer in the doc)[https://docs.e3nn.org/en/stable/guide/transformer.html#transformer-guide]?

**Reply by thu-wangz17** [reply]
Yes, we are talking the same thing. Since in the original paper, it said that  "The linear embedding matrices $W_Q^{lk}$ and $W^{lk}_K(xj − xi) are of TFN type (c.f. Eq. (8)). "  Thus the forms of the $W_Q$ and $W_k$ seem to be the same except the weights. I'm confused that why is the $W_Q$ implemented with `o3.Linear` module rather than the module of $W_K$, i.e. `o3.FullyConnectedTensorProduct`?

**Reply by mariogeiger** [reply]
`o3.FullyConnectedTensorProduct` is bi-linear, it multiplies two inputs together. 

`o3.Linear` is just linear. 

Queries are nodes properties but Values and Keys are edge properties. 

To make edge properties we use the node's source features and the spherical harmonics. We multiply the two with a bi-linear operation. 

To make the queries we only use the node features so we use a linear operation.

**Reply by thu-wangz17** [reply]
That's very clear. Thank you very much.
